source/common/upstream/cluster_manager_impl.cc

```diff
@@ -479,6 +479,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
       // Otherwise, applyUpdates() will fire with a dangling cluster reference.
       updates_map_.erase(cluster_name);
 
+      // If the active cluster has priority sets and the warming cluster does not have them, it
+      // means that onConfigUpdate is triggered for EDS type of cluster is triggered with no
+      // reference to this cluster. In such cases, copy the active cluster priority set to the
+      // warming cluster otherwise the active cluster hosts will be cleared after warming is
+      // completed. See https://github.com/envoyproxy/envoy/issues/5168 for more context.
+      const auto active_it = active_clusters_.find(cluster_name);
+      if (active_it != active_clusters_.end()) {
+        auto& active_cluster_entry = *active_it->second;
+        if (active_cluster_entry.cluster_->prioritySet().hostSetsPerPriority().size() != 0 &&
+            (cluster_entry.cluster_->prioritySet().hostSetsPerPriority().size() == 0 ||
+             (cluster_entry.cluster_->prioritySet().hostSetsPerPriority().size() == 1 &&
+              cluster_entry.cluster_->prioritySet().hostSetsPerPriority()[0]->hosts().size() ==
+                  0))) {
+          ENVOY_LOG(
+              debug,
+              "copying host set from active cluster {} as warming cluster does not have host set",
+              cluster_name);
+
+          const auto& host_sets =
+              active_cluster_entry.cluster_->prioritySet().hostSetsPerPriority();
+          for (size_t priority = 0; priority < host_sets.size(); ++priority) {
+            const auto& host_set = host_sets[priority];
+            HostVectorConstSharedPtr hosts_copy(new HostVector(host_set->hosts()));
```

Comment:
I tried, but did not know to do it - I think it may be related to this comment https://github.com/envoyproxy/envoy/blob/1c9d6d4359aa4e1a1d0e791b388e89e6bb3983b1/source/common/upstream/cluster_manager_impl.cc#L655. LMK if you have better idea on how to do it?

Reply:
Ah, the issue is that `partitionHosts` takes a vector of const pointers, while hosts() returns non-const pointers. In that case, mind adding a TODO here as well? I think it would make sense to look into optimizing away these copies at a later time.

Reply:
added the TODO